### ISSUE TITLE
setState not working on slideChange of carousel inside amp-render

### ISSUE THREAD

**Issue by gskaushal92** [ampproject/amphtml#35484]
### Description

I am trying to setstate on slide change inside amp-render. But it is not reflecting.

### Reproduction Steps

1. I am loading data inside amp-carousel using amp-render.
2. I am binding setState on carousel slide change
3. It is not working.

is this is a known issue. or is there any workaround.

**Comment by gskaushal92** [comment]
Anybody has any update on this?

**Comment by gskaushal92** [comment]
any update on this issue?

**Comment by gskaushal92** [comment]
@caroqliu  setState is not working inside amp-render. Is this known issue?

**Comment by caroqliu** [comment]
@dmanek Can you take a look at this? I've created a [minimal sample in the playground here](https://playground.amp.dev/#share=PCFET0NUWVBFIGh0bWw+CjxodG1sIOKaoT4KICA8aGVhZD4KICAgIDxtZXRhIGNoYXJzZXQ9InV0Zi04IiAvPgogICAgPHRpdGxlPk15IEFNUCBQYWdlPC90aXRsZT4KICAgIDxsaW5rIHJlbD0iY2Fub25pY2FsIiBocmVmPSJzZWxmLmh0bWwiIC8+CiAgICA8bWV0YSBuYW1lPSJ2aWV3cG9ydCIgY29udGVudD0id2lkdGg9ZGV2aWNlLXdpZHRoIiAvPgogICAgPHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7LW1vei1hbmltYXRpb246LWFtcC1zdGFydCA4cyBzdGVwcygxLGVuZCkgMHMgMSBub3JtYWwgYm90aDstbXMtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7YW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGh9QC13ZWJraXQta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUAtbW96LWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW1zLWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW8ta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUBrZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19PC9zdHlsZT48bm9zY3JpcHQ+PHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOm5vbmU7LW1vei1hbmltYXRpb246bm9uZTstbXMtYW5pbWF0aW9uOm5vbmU7YW5pbWF0aW9uOm5vbmV9PC9zdHlsZT48L25vc2NyaXB0PgogICAgPHNjcmlwdCBhc3luYyBzcmM9Imh0dHBzOi8vY2RuLmFtcHByb2plY3Qub3JnL3YwLmpzIj48L3NjcmlwdD4KICAgIDxzY3JpcHQgYXN5bmMgY3VzdG9tLWVsZW1lbnQ9ImFtcC1jYXJvdXNlbCIgc3JjPSJodHRwczovL2Nkbi5hbXBwcm9qZWN0Lm9yZy92MC9hbXAtY2Fyb3VzZWwtMC4yLmpzIj48L3NjcmlwdD4KICAgIDxzY3JpcHQgYXN5bmMgY3VzdG9tLWVsZW1lbnQ9ImFtcC1yZW5kZXIiIHNyYz0iaHR0cHM6Ly9jZG4uYW1wcHJvamVjdC5vcmcvdjAvYW1wLXJlbmRlci0xLjAuanMiPjwvc2NyaXB0PgogICAgPHNjcmlwdCBhc3luYyBjdXN0b20tdGVtcGxhdGU9ImFtcC1tdXN0YWNoZSIgc3JjPSJodHRwczovL2Nkbi5hbXBwcm9qZWN0Lm9yZy92MC9hbXAtbXVzdGFjaGUtMC4yLmpzIj48L3NjcmlwdD4KICAgIDwhLS0gPHNjcmlwdCBhc3luYyBjdXN0b20tZWxlbWVudD0iYW1wLWJpbmQiIHNyYz0iaHR0cHM6Ly9jZG4uYW1wcHJvamVjdC5vcmcvdjAvYW1wLWJpbmQtMC4xLmpzIj48L3NjcmlwdD4gLS0+CiAgICA8c3R5bGUgYW1wLWN1c3RvbT4KICAgICAgaDEgewogICAgICAgIG1hcmdpbjogMXJlbTsKICAgICAgfQogICAgICAuc2xpZGUgewogICAgICAgIGJhY2tncm91bmQtY29sb3I6IGdyYXk7CiAgICAgICAgY29sb3I6IHdoaXRlOwogICAgICAgIGRpc3BsYXk6IGZsZXg7CiAgICAgICAganVzdGlmeS1jb250ZW50OiBjZW50ZXI7CiAgICAgICAgYWxpZ24taXRlbXM6IGNlbnRlcjsKICAgICAgICBoZWlnaHQ6IDEwMCU7CiAgICAgIH0KICAgIDwvc3R5bGU+CiAgPC9oZWFkPgogIDxib2R5PgogICAgPHA+CiAgICAgIENhcm91c2VsIHN0YXRlOiA8c3BhbiBbdGV4dF09ImN1cnJlbnRTbGlkZSI+bnVsbDwvc3Bhbj4KICAgIDwvcD4KICAgIDxhbXAtcmVuZGVyIGlkPSJteS1jb21wb25lbnQiIHNyYz0iaHR0cHM6Ly9hbXAuZGV2L3N0YXRpYy9pbmxpbmUtZXhhbXBsZXMvZGF0YS9hbXAtbGlzdC1kYXRhLmpzb24iIGhlaWdodD0iMzAwIj4KICAgICAgPHRlbXBsYXRlIHR5cGU9ImFtcC1tdXN0YWNoZSI+CiAgICAgICAgPGFtcC1jYXJvdXNlbCB0eXBlPSJzbGlkZXMiIGhlaWdodD0iMzAwIiBvbj0ic2xpZGVDaGFuZ2U6QU1QLnNldFN0YXRlKHtjdXJyZW50U2xpZGU6IGV2ZW50LmluZGV4fSkiPgogICAgICAgICAge3sjaXRlbXN9fQogICAgICAgICAgPGFtcC1pbWcgc3JjPSJ7e2ltYWdlVXJsfX0iIGxheW91dD0iZmxleC1pdGVtIj48L2FtcC1pbWc+CiAgICAgICAgICB7ey9pdGVtc319CiAgICAgICAgPC9hbXAtY2Fyb3VzZWw+CiAgICAgIDwvdGVtcGxhdGU+CiAgICA8L2FtcC1yZW5kZXI+CiAgPC9ib2R5Pgo8L2h0bWw+Cg==) of what I understand of the issue.

Here is what I'm observing:
- `amp-render` renders the `amp-carousel` from the data. There is an `on` action to `setState` given the `onSlideChange` event. 
- Advance on the carousel. You'll notice that nothing changes on the document. This is because the console displays the following error:
  ```
  Uncaught (in promise) Error: AMP-BIND is not installed.​​​
    at eb (VM229 v0.js:2)
    at Db.g.assert (VM229 v0.js:2)
    at K (VM229 v0.js:2)
    at VM229 v0.js:5
  ```
- installing `amp-bind` on the document causes the `amp-carousel` not to render (?)

**Comment by gskaushal92** [comment]
If I move the binding inside template then amp-carousel will render. But again on slide Change setState will not be set.

`<!DOCTYPE html>
<html ⚡>
  <head>
    <meta charset="utf-8" />
    <title>My AMP Page</title>
    <link rel="canonical" href="self.html" />
    <meta name="viewport" content="width=device-width" />
    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
    <script async src="https://cdn.ampproject.org/v0.js"></script>
    <script async custom-element="amp-carousel" src="https://cdn.ampproject.org/v0/amp-carousel-0.2.js"></script>
    <script async custom-element="amp-render" src="https://cdn.ampproject.org/v0/amp-render-1.0.js"></script>
    <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
    
    <style amp-custom>
      h1 {
        margin: 1rem;
      }
      .slide {
        background-color: gray;
        color: white;
        display: flex;
        justify-content: center;
        align-items: center;
        height: 100%;
      }
    </style>
  </head>
  <body>
    <amp-render id="my-component" src="https://amp.dev/static/inline-examples/data/amp-list-data.json" height="300">
      <template type="amp-mustache">
        <p>
      Carousel state: <span [text]="currentSlide">null</span>
    </p>
        <amp-carousel type="slides" height="300" on="slideChange:AMP.setState({currentSlide: event.index})">
          {{#items}}
          <amp-img src="{{imageUrl}}" layout="flex-item"></amp-img>
          {{/items}}
        </amp-carousel>
      </template>
    </amp-render>
  </body>
</html>
`

**Comment by gskaushal92** [comment]
@caroqliu  So I hope this is a known issue and will be taken care by your team?

**Comment by dmanek** [comment]
@gskaushal92 thanks for reporting this issue. Will investigate and update when I have more.

**Comment by gskaushal92** [comment]
@dmanek  any idea when can we get the fix for this issue?

**Comment by dmanek** [comment]
@gskaushal92 I have the PR out for review. Once it's merged, it should be on rolled out in ~2 weeks.

**Comment by dmanek** [comment]
Closing this issue since it's been fixed by PR #35449.

This issue is caused due to amp-img within amp-carousel having `display: flex` causing it not to display since the container div's height is 0. In PR #35449 we preserve the wrapper element which allows amp-img to have height > 0.
